### DISCUSSION TITLE
Proposal: UART over USB for Logging and Debugging

### DISCUSSION THREAD

**Discussion by KozhinovAlexander** [CANnectivity/cannectivity#36] [Ideas]
Adding UART over USB using the Console over USB CDC ACM protocol in Zephyr RTOS is a valuable feature for compact, sparse-designed boards especially without on-board debugger.

### Advantages

1. **Space-saving**: Utilizes USB instead of dedicated UART pins, ideal for boards with limited space.
2. **Existing Infrastructure**: USB is already required for this project

### Implementation Considerations

- **KConfig Option**: Offer this feature as a compile-time option to conserve resources on memory-constrained boards.
- **Minimal Overhead**: The implementation adds little overhead to the existing USB stack.
- **Compatibility**: Integrate smoothly with Zephyr’s logging and shell subsystems.

### Benefits

1. **Simplified Debugging**: Easy access to console output without extra hardware.
2. **Flexibility**: Useful for boards lacking exposed UART pins or with limited GPIO.

@henrikbrixandersen what do you think or dou you have it already?

UPD: used perplexity.ai to make this text "better"

**Comment by henrikbrixandersen** [comment]
This is already supported by upstream Zephyr, e.g. by using the CDC-ACM snippet when building:

`west build -b usb2canfdv1/stm32g0b1xx ../custom/cannectivity/app/ -S cdc-acm-console`

This has issues, though, when it comes to e.g. Linux driver compatibility. The Linux kernel gs_usb driver will attempt to use interface 0, which may now be assigned to the CDC-ACM class. So now the CANnectivty firmware logging/console over CDC-ACM works, but the gs_usb driver will not recognise the device.

I'd recommend using a RAM-based console like the Segger J-Link RTT console for debugging instead.

**Reply by KozhinovAlexander** [reply]
Thanks - I see. I've used zephyr cdc-acm yesterday without gs_usb. Need to see behavior on updated gs_usb driver.

Regarding segger rtt can't find licensing/pricing considerations. It would be nice to have none of them, cause segger is not cheap.

**Reply by henrikbrixandersen** [reply]
While RTT was initially developed by Segger, it is not limited to their tooling. See for instance https://openocd.org/doc/html/General-Commands.html#Real-Time-Transfer-_0028RTT_0029

You can also opt to use the generic Zephyr `CONFIG_RAM_CONSOLE` which can be read via any debugger.

**Reply by KozhinovAlexander** [reply]
@henrikbrixandersen Thank you. I've tried rtt - it is awesome :)
